Add input source, image artwork to the TouchBar, fix battery icon, add estimated charge/remaining time + additional fixes + better Big Sur support

konstantintuev opened the pull request:
**Describe the Pull request**
Add the icon of the currently used input source and change it when the input source is changed 


**Screenshots**
<img width="1004" alt="Touch Bar Shot 2020-06-21 at 21 10 48" src="https://user-images.githubusercontent.com/30181495/85231901-bdc61b80-b403-11ea-8707-595e86d1b999.png">
<img width="1004" alt="Touch Bar Shot 2020-06-21 at 21 10 36" src="https://user-images.githubusercontent.com/30181495/85231903-bef74880-b403-11ea-8c1f-3b486e078161.png">
<img width="1004" alt="Touch Bar Shot 2020-06-22 at 13 25 44" src="https://user-images.githubusercontent.com/30181495/85277340-f6a9d300-b48b-11ea-8bb6-40e884cf4d9e.png">
<img width="1004" alt="Touch Bar Shot 2020-06-22 at 13 26 57" src="https://user-images.githubusercontent.com/30181495/85277437-1640fb80-b48c-11ea-9dac-f04373a457a5.png">




**Pull request type**
Keep only the option that better matches your pull request.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)


**Tested?**
I compiled and ran from and outside of XCode.

I also made changes to the Preferences which are NOT committed because `PreferencePane.Identifier` was not found and I replaced it and all other references of `Identifier` with `Preferences.PaneIdentifier`


*Test Configuration*:
* Hardware:      MacBook Pro (16-inch, 2019)
* macOS Version: 10.15.5
* Xcode version: 11.5


**Checklist**
Use this list to keep track of your progress:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works

Egwor commented on Pock/Widgets/NowPlaying/NowPlayingHelper.swift:
```diff
@@ -70,19 +70,200 @@ class NowPlayingHelper {
+                // fallback to network artwork after 20 seconds
+                DispatchQueue.main.async {
+                    if (self.timesLeftTryUpdatingMediaContentManually <= 0) {
+                        // we have tried getting the artwork with official api manuall but it didn't work, get it from the iTunes API
```
```suggestion
                        // we have tried getting the artwork with official api manually but it didn't work, get it from the iTunes API
```

konstantintuev replied:
Thanks!

Egwor commented on Pock/Widgets/NowPlaying/NowPlayingHelper.swift:
```diff
@@ -70,19 +70,200 @@ class NowPlayingHelper {
+                    if (self.timesLeftTryUpdatingMediaContentManually <= 0) {
+                        // we have tried getting the artwork with official api manuall but it didn't work, get it from the iTunes API
+                        self.timesLeftTryUpdatingMediaContentManually = 2
+                        self.albumArtFallbackTimer = Timer.scheduledTimer(timeInterval: 20.0, target: self, selector: #selector(NowPlayingHelper.fireTimerIconFailed), userInfo: nil, repeats: false)
```
You could simplify the code by setting the fallback artwork first, and then once the iTunes API returns the image could be updated.

konstantintuev replied:
The iTunes API is the least good solution because it can make mistakes when searching by artist and song name.
The native player API sometimes doesn't automatically notify the program when the artwork has been loaded, that's why I check for it manually first after 1.5 secs and then after 8 secs. When this fails the iTunes API is used after 5 secs.

konstantintuev commented on Pock/Widgets/NowPlaying/NowPlayingHelper.swift:
```diff
@@ -74,6 +74,73 @@ class NowPlayingHelper {
+        }
+        
+        // Start fetching artwork
+        artworkAPITask = URLSession.fetchJSON(fromURL: URL(string: "https://itunes.apple.com/search?term=\(search)&entity=song&limit=1")!) { (data, json, error) in
```
I have made it check with the iTunes API only once no matter what because when playing a generic YouTube video for example.
The last used iTunes API image is cached by title and raw data.
Before my last changes the timer loop wasn't working at all but I fixed it. The timer loop is called only 3 times.

konstantintuev commented on Pock/Widgets/NowPlaying/NowPlayingHelper.swift:
```diff
@@ -74,6 +74,73 @@ class NowPlayingHelper {
+        
+        // Start fetching artwork
+        artworkAPITask = URLSession.fetchJSON(fromURL: URL(string: "https://itunes.apple.com/search?term=\(search)&entity=song&limit=1")!) { (data, json, error) in
+            if error != nil {
```
I check the array length, if data is nil or if there is an error.

Egwor replied:
missed that, thanks

konstantintuev commented on Pock/Widgets/NowPlaying/NowPlayingHelper.swift:
```diff
@@ -74,6 +74,73 @@ class NowPlayingHelper {
+        }
+        
+        
+        // Destroy tasks, if any was already busy
```
Thanks!

L1cardo commented:
This is cool!

konstantintuev commented:
I have been using my build of Pock for quite some time and I have had no crashes, even ones which were happening on stock Pock.

L1cardo commented:
Album image works fine for me, but the input source icon is not showing.
@pigigaldi is developing PockKit now, which allows you to develop your own widget, so this PR won't be merged in a short time, please have patience.
<img width="1004" alt="触控栏快照2020-07-22 下午1 38 38" src="https://user-images.githubusercontent.com/33802186/88138631-a6f03000-cc20-11ea-8915-e219c3ec6185.png">

konstantintuev commented:
Fixed, input sources like Chinese or Japanese are black icons with transparent character.
When such input source is selected it needs to be tinted white.
It was an easy fix because these kind of icons are loaded differently than the flag icons.

I have no problem waiting, just providing some feedback.

L1cardo commented:
@konstantintuev I changed my input method to English(America), Germany(German), still is not working

konstantintuev commented:
There were some issues with black icons like ABC - QWERTZ which couldn't be detected trivially and tinted.
I added a workaround which records 16 points of the image positioned in a grid to get the dominant color and save the number of black colours found.

Help about the way the way the MenuBar finds black/white images and colours them would be appreciated.

The issue you are showing seems to be Big Sur specific, when I have some more time I will install a VM and try checking it out.

konstantintuev commented:
Also thank you for taking the time to test!

Hamitamaru commented:
Any chance this can be updated to show a cover image placed into the artwork attribute of MediaMetadata https://developer.mozilla.org/en-US/docs/Web/API/MediaMetadata/MediaMetadata . When I play stuff in chrome and set the title and artist attributes it shows in pock.

konstantintuev commented:
I worked on fixing the battery icon on BigSur (completely changed everything because the way it was updated is not compatible with Big Sur - refreshing the UI every 1 second, most of the code is from [Apple Juice](https://github.com/raphaelhanneken/apple-juice)), added a limit to clock refreshes, fixed the Clock and Language item black out, fixed delay when updating the battery status (because now Pock attaches to the system service instead of polling), added an option to the preferences to show estimated time needed for battery charge/discharge instead of a percentage, added an option to the preferences to load album art for the currently playing item, worked around NSStackView issues with the constraints on Big Sur

Here is how it looks:
<img width="1004" alt="Touch Bar Shot 2020-11-18 at 23 16 31" src="https://user-images.githubusercontent.com/30181495/99589913-541f5d00-29f5-11eb-814c-5048ade8d9e1.png">
<img width="1004" alt="Touch Bar Shot 2020-11-18 at 23 24 43" src="https://user-images.githubusercontent.com/30181495/99589929-58e41100-29f5-11eb-8698-7b1f027909d4.png">

konstantintuev commented:
You can test and debug using the build here: https://github.com/konstantintuev/Pock/releases/tag/v.0.7.2%2Bchanges
Note that it's debug build and will consume marginally more more battery than a release build (~0.5-1.0% more).

konstantintuev commented:
I am starting to maintain my own fork of Pock - [PockV2](https://github.com/konstantintuev/Pock).
Check it out and head to the [Releases](https://github.com/konstantintuev/Pock/releases) tab to get the newest version!

aldotroiano reviewed (approved):
Working commit. Fixes battery issue
